### PR TITLE
relay: Refactor hosts/stats interface

### PR DESCRIPTION
**Pull request by prashantv**
This refactor does not make any logical changes to the code -- although
it does allow for relay to now forward frames across channels by
returning a peer from a separate channel.

The new interface combines the peer selection and stats interface into a
single RelayCall interface that returns a *Peer rather than a
relay.Peer.

Some advantages of this approach:
 - Avoids a map lookup for converting host:port to *Peer
 - Allows returning peers from separate channels
   This allows our Hyperbahn emulation code to significantly reduce the
   number of channels we will need.
 - No more passing around Zone/Pool in tchannel which does not need to
   know about these concepts.

### CODE REVIEW

File: relay_api.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	Start(relay.CallFrame, relay.Conn) (RelayCall, error)
+}
+
+// RelayCall abstracts away peer selection, stats, and any other business
+// logic from the underlying relay implementation. A RelayCall may not
+// have a destination if there was an error during peer selection
+// (which should be returned from start).
+type RelayCall interface {
+	// Destination returns the selected peer (if there was no error from Start).
+	Destination() (peer *Peer, ok bool)
```
Comment (akshayjshah) [inline relay_api.go]: Why do we need the additional boolean? Is there a situation where `peer == nil` but `ok == true`?

File: benchmark/real_relay.go [hunk 2]
```diff
@@ -34,14 +36,14 @@ type fixedHosts struct {
 	pickI atomic.Int32
 }
 
-func (fh *fixedHosts) Get(call relay.CallFrame, _ relay.Conn) (relay.Peer, error) {
-	peers := fh.hosts[string(call.Service())]
+func (fh *fixedHosts) Get(cf relay.CallFrame, conn relay.Conn) (string, error) {
```
Comment (prashantv) [inline benchmark/real_relay.go]: ✅

File: connection.go [hunk 3]
```diff
@@ -309,7 +309,7 @@ func (ch *Channel) newConnection(conn net.Conn, outboundHP string, initialState
 	c.inbound.onAdded = c.onExchangeAdded
 	c.outbound.onAdded = c.onExchangeAdded
 
-	if ch.relayHosts != nil {
+	if ch.relayHost != nil {
```
Comment (prashantv) [inline connection.go]: ✅

File: relay/noop.go [hunk 4]
```diff
@@ -19,21 +19,3 @@
...
-func (n noopStats) Begin(_ CallFrame) CallStats {
-	return noopCallStats{}
-}
-
-type noopCallStats struct{}
-
-func (n noopCallStats) Succeeded()      {}
-func (n noopCallStats) Failed(_ string) {}
-func (n noopCallStats) SetPeer(_ Peer)  {}
-func (n noopCallStats) End()            {}
```
Comment (prashantv) [inline relay/noop.go]: ✅ 

wasn't sure whether i'd need noop implementations, but we can add them later if required.

File: relay_api.go (hunk 1 above)
Comment (prashantv) [inline relay_api.go]: Nope, if `ok`, then `peer != nil`. Just added this to make it more explicit that `peer` can be `nil`.

**Pull request merged**